rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,678 @@
+
+   <details><summary>Arguments</summary>
+
+   - (+) Formally decouples the Nix language version from the Nix version
+     - (+) The Nix language is supposed to change much less often than the rest of Nix
+     - (-) There are two version numbers to keep track of
+     - (-) Makes more evident that the Nix language is a distinct architectural component of the Nix ecosystem
+   - (+) It's currently handled that way, no change needed apart from documentation
+        - See [`builtins.langVersion`] (currently undocumented)
+   - (+) Simple and unambiguous
+   - (+) Concise, even in the long term, since the language is supposed to change very rarely
+
+   [`builtins.langVersion`]: https://github.com/NixOS/nix/blob/26c7602c390f8c511f326785b570918b2f468892/src/libexpr/primops.cc#L3952-L3957
+
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - [Calendar Versioning](https://calver.org/)
+       - (+) Provides information on when changes happened
+           - (-) This is not needed because only compatibility information is needed
+       - (-) Requires a minimum amount of characters
+           - This may be relevant depending on where it has to be encoded
+       - (+) Restricting to only the year would force language changes to be rare
+           - (+) This would allow obvious synchronisation points with Nixpkgs releases
+           - (-) It may be too much policy encoded in a mechanism
+   - [Semantic Versioning](https://semver.org/)
+       - (+) Can distinguish additions from other changes
+           - (-) This is not needed for our use case, since any addition to an expression will break for older evaluators even if the major version matches
+       - (-) Requires more characters to account for the added expressiveness
+           - This may be relevant depending on where it has to be encoded
+   - Use version numbers of Nix stable releases for specifying the version of the Nix language
+      - (+) More obvious to see for users what the current Nix version is rather than `builtins.langVersion`
+      - (-) Would tie alternative Nix language evaluators to the rest of Nix
+        - (-) One can add a command line option such that it is not more effort than `nix --version`
+          - (+) That requires adding another built-in to the public API
+        - (-) Using a language feature requires an additional steps from users to determine the current version
+            - (-) Requires adding another command line option to the public API
+      - (+) The Nix language version is decoupled Nix version numbering
+        - (+) It changes less often than the Nix version
+          - (-) That was probably due to making changes being so hard
+            - (+) The language changing slowly is a desirable property for wider adoption
+        - (-) There are two version numbers to keep track of
+
+   </details>
+
+1. The language version for Nix expressions is denoted in special syntax, at the beginning of a parse unit.
```

Comment:
```suggestion
1. The language version for Nix expressions is denoted in special syntax, which makes up the first statement of a file.
```

Reply:
I'm not aware we even have a notion of "statement".

Reply:
True, the statement is wrong, "top-level expression" would probably be more precise. Yes the next sentence needs to be removed, I just didn't get GitHub to render that correctly.

```suggestion
1. The language version for Nix expressions is denoted in special syntax, which makes up the first expression in the file (or in other words, the top-level node of the AST).
```